### DISCUSSION TITLE
ARE THERE ALTERNATIVES TO ADDRESS-KEYED MAP STORAGE ON ALGO

### DISCUSSION THREAD

**Discussion by kcelestinomaria** [reach-sh/reach-lang#1232] [Q&A]
Hello,

Since you can't create non-address Map keys for storage on ALGO, are there any alternatives to this? If not, what's the progress at with with the ALGO-Reach collaboration at making it possible with having non-address types as keys to Maps on Reach Smart Contracts? Are we going to see an upgrade anytime soon? This feature not being there messes up with ensuring persistent storage of data. There is also another problem with Maps on Reach in that once a user adds new Mapping data(specifically a new key -> value pair), it overwrites the previous key-value pair, this makes it hard to search up past data on Reach, access it and use it in future needs.

Thanks.

**Comment by jeapostrophe** [comment]
You can see the progress that Algorand is making on the feature we'll use to implement this part of Reach on Algorand --- https://github.com/algorand/go-algorand/pull/4149

They announced at AVM in Austin at the beginning of June that it will be released this summer.

For your second question, this is what it means to have a map. There can only be one 41st state (Montana) and there can only be one height of "Jay McCarthy". If you want more than one, then you really have a more complex key, such as "Jay McCarthy at 15" and "Jay McCarthy at 9". Of course, this means that you need to wait until more keys are added.

A non-specific and obtuse workaround is to rearrange your application so that each map cell is like a different application (contract) and use some external mapping mechanism that your clients reconstruct.
